src/Facility.LanguageServer/FsdDefinitionUtility.cs

```diff
@@ -17,20 +17,81 @@ public static ServiceMemberInfo GetMemberReferencedAtPosition(this ServiceInfo s
 				select service.FindMember(name)).FirstOrDefault();
 		}
 
+		public static IEnumerable<ServicePart> GetReferencedServicePartsAtPosition(this ServiceInfo service, Position requestPosition)
+		{
+			var members = service.GetDescendants().OfType<ServiceMemberInfo>().ToList().AsReadOnly();
+
+			// memberNameAtPosition will be null if the cursor is not on a member name.
+			var memberNameAtCursor = members
+				.Select(member =>
+				{
+					var part = member.GetPart(ServicePartKind.Name);
+					var name = member.Name;
+
+					return new { part, name };
+				})
+				.Where(x => x.part != null && requestPosition >= x.part.Position && requestPosition < x.part.EndPosition)
+				.Select(x => x.name)
+				.FirstOrDefault();
+
+			var fields = service.GetDescendants().OfType<ServiceFieldInfo>().ToList().AsReadOnly();
+
+			// fieldTypeNameAtPosition will be null if the cursor is not on a field type name.
+			var fieldTypeNameAtCursor = fields
+				.Select(field =>
+				{
+					var part = field.GetPart(ServicePartKind.TypeName);
+					var typeName = GetFieldTypeName(field);
+
+					return new { part, typeName };
+				})
+				.Where(x => x.part != null && requestPosition >= x.part.Position && requestPosition < x.part.EndPosition)
+				.Select(x => x.typeName)
+				.FirstOrDefault();
+
+			return fields
+				.Select(field =>
+				{
+					var part = field.GetPart(ServicePartKind.TypeName);
+					var typeName = GetFieldTypeName(field);
+
+					var type = service.GetFieldType(field);
+					var memberTypeName = type?.GetMemberTypeName();
+
+					return new { part, memberTypeName, typeName };
+				})
+				.Where(x => x.part != null && ((memberNameAtCursor != null && x.memberTypeName == memberNameAtCursor) || x.typeName == fieldTypeNameAtCursor))
+				.Select(x => x.part);
+		}
+
 		private static string GetMemberTypeName(this ServiceTypeInfo type)
 		{
 			switch (type.Kind)
 			{
 				case ServiceTypeKind.Dto:
 					return type.Dto!.Name;
+				case ServiceTypeKind.ExternalDto:
+					return type.ExternalDto!.Name;
 				case ServiceTypeKind.Enum:
 					return type.Enum!.Name;
+				case ServiceTypeKind.ExternalEnum:
+					return type.ExternalEnum!.Name;
 				case ServiceTypeKind.Array:
 				case ServiceTypeKind.Result:
 				case ServiceTypeKind.Map:
 					return type.ValueType.GetMemberTypeName();
 			}
 			return null;
 		}
+
+		private static string GetFieldTypeName(this ServiceFieldInfo field)
+		{
+			return field.TypeName
+				.Replace("[]", "", StringComparison.Ordinal)
+				.Replace("map<", "", StringComparison.Ordinal)
+				.Replace("nullable<", "", StringComparison.Ordinal)
+				.Replace("result<", "", StringComparison.Ordinal)
+				.Replace(">", "", StringComparison.Ordinal);
+		}
```

Comment:
I didn't see a way to handle this from the parser. Is there a better way that exists? Would this be worth adding awareness of value types for fields to the parser if that doesn't exist already?

Reply:
I'm probably not understanding, but can you get that information from `service.GetFieldType(field)`?